stan/math/prim/functor/apply_scalar_unary.hpp

```diff
@@ -59,15 +58,82 @@ struct apply_scalar_unary<F, T, require_eigen_t<T>> {
    * Return the result of applying the function defined by the
    * template parameter F to the specified matrix argument.
    *
+   * @tparam DenseMat A type derived from `Eigen::DenseBase`.
    * @param x Matrix to which operation is applied.
    * @return Componentwise application of the function specified
    * by F to the specified matrix.
    */
-  static inline auto apply(const T& x) {
+  template <typename DenseMat, require_eigen_dense_base_t<DenseMat>* = nullptr>
+  static inline auto apply(const DenseMat& x) {
     return x.unaryExpr(
         [](scalar_t x) { return apply_scalar_unary<F, scalar_t>::apply(x); });
   }
 
+  /**
+   * Special case for `ApplyZero` set to true, returning a full sparse matrix.
+   * Return the result of applying the function defined by the template
+   * parameter F to the specified matrix argument.
+   *
+   * @param SparseMat A type derived from `Eigen::SparseMatrixBase`
+   * @tparam NonZeroZero Shortcut trick for using class template for deduction,
+   * should not be set manually.
+   * @param x Matrix to which operation is applied.
+   * @return Componentwise application of the function specified
+   * by F to the specified matrix.
+   */
+  template <typename SparseMat, bool NonZeroZero = ApplyZero,
+            require_t<bool_constant<NonZeroZero>>* = nullptr,
+            require_eigen_sparse_base_t<SparseMat>* = nullptr>
+  static inline auto apply(const SparseMat& x) {
+    using val_t = value_type_t<SparseMat>;
+    using triplet_t = Eigen::Triplet<val_t>;
+    auto zeroed_val = apply_scalar_unary<F, scalar_t>::apply(val_t(0.0));
+    const auto x_size = x.size();
+    std::vector<triplet_t> triplet_list(x_size, triplet_t(0, 0, zeroed_val));
+    for (Eigen::Index i = 0; i < x.rows(); ++i) {
+      for (Eigen::Index j = 0; j < x.cols(); ++j) {
+        // Column major order
+        triplet_list[i * x.cols() + j] = triplet_t(i, j, zeroed_val);
+      }
+    }
+    for (Eigen::Index k = 0; k < x.outerSize(); ++k) {
+      for (typename SparseMat::InnerIterator it(x, k); it; ++it) {
+        triplet_list[it.row() * x.cols() + it.col()]
```

Comment:
Rather than iterating over each element, would it be more performant to Map the values as a vector and call the vectorised implementation? For example:
```cpp
Map<Eigen::VectorXd> mapSparse(x.valuePtr(), x.nonZeros());

apply_scalar_unary<F, Eigen::VectorXd>::apply(mapSparse);
```

Reply:
It would def be better, though I avoided it because I was a bit confused about the compressed sparse matrix scheme Eigen uses. If you look at their docs [here](https://eigen.tuxfamily.org/dox/group__TutorialSparse.html) under 'The SparseMatrix class' they sometimes allow empty values so its easier to write to the sparse matrix like

```
Values:	22	7	_	3	5	14	_	_	1	_	17	8
InnerIndices:	1	2	_	0	2	4	_	_	2	_	1	4
```

I was worried that directly grabbing the values could lead to some bad things since there may be places in the value pointer that are not initialized. Another issue is that our `arena_matrix` scheme assumes that if you are passing it an Eigen map that the memory has already been allocated elsewhere for use in the reverse mode pass, which may not explicitly be true for sparse matrices.

So I agree what your saying here is def faster, but imo for this first pass I'd rather do the safer path atm. Once we have everything working at the language level and know more about our constraints and assumptions we can make then we can come back and do these kinds of optimizations.

Reply:
Hmm, okay now I'm thinking about this more and I think that at the stan language level we can make it so that you can only assign to elements that are already non-zero. So I wonder if what we can do here is just assume that we are always working with compressed matrices, since then we don't have to worry about the uninitialized values and can do the fast thing safely